Convert client_channel.c to use a combiner lock

ctiller opened the pull request:
Next steps are to expand the radius of this lock to contain lb_policy, resolver.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/grpc/grpc/9660)
<!-- Reviewable:end -->

ctiller commented:
Now built on #9666

markdroth commented:
On the whole, this looks really good.  Most of my comments are minor; my only significant concern is the use of the transport_private data in a couple of places.

Note that I have not reviewed any of the pollset or completion_queue changes, since I assume that Sree will do that in #9666.  That having been said, I'm curious as to why you merged that in to this PR, since the two changes seem orthogonal.  Am I missing some connection between them?

---

Reviewed 19 of 19 files at r1.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[src/core/ext/client_channel/client_channel.c, line 427 at r1](https://reviewable.io:443/reviews/grpc/grpc/9660#-KcczbwpRpQCU_BjKzgU:-KcczbwpRpQCU_BjKzgV:b-d67ifc) ([raw file](https://github.com/grpc/grpc/blob/3fc3408cd33eb2e221863e0881ec82eb8e21414f/src/core/ext/client_channel/client_channel.c#L427)):*
> ```C
>                                          grpc_error *error_ignored) {
>   grpc_transport_op *op = arg;
>   grpc_channel_element *elem = op->transport_private.args[0];
> ```

It seems a bit messy to use transport private data here, since this is not transport code.  Could we instead create some sort of local data structure to pass in both op and elem via args?

---

*[src/core/ext/client_channel/client_channel.c, line 876 at r1](https://reviewable.io:443/reviews/grpc/grpc/9660#-Kcd0XSCchVKu8_DRk-4:-Kcd0XSCchVKu8_DRk-5:biguv80) ([raw file](https://github.com/grpc/grpc/blob/3fc3408cd33eb2e221863e0881ec82eb8e21414f/src/core/ext/client_channel/client_channel.c#L876)):*
> ```C
>                                                 grpc_error *error_ignored) {
>   grpc_transport_stream_op *op = arg;
>   grpc_call_element *elem = op->transport_private.args[0];
> ```

Same comment here as above -- I'd prefer to use some other mechanism to pass both op and elem in via args.

---

*[src/core/ext/client_channel/client_channel.c, line 881 at r1](https://reviewable.io:443/reviews/grpc/grpc/9660#-Kcd1xz3nTvQgyjCQKwN:-Kcd1xz3nTvQgyjCQKwO:bbomoat) ([raw file](https://github.com/grpc/grpc/blob/3fc3408cd33eb2e221863e0881ec82eb8e21414f/src/core/ext/client_channel/client_channel.c#L881)):*
> ```C
>   grpc_subchannel_call *call;
> 
> retry:
> ```

Now that the logic here is a bit simpler, would it make sense to use a loop instead of a goto for this?

Feel free to push back on this -- I'm honestly not sure whether or not this would improve readability.  I do generally prefer loops to gotos, but in this particular case, it might be that "goto retry" and "goto done" are a bit more self-documenting than "continue" and "break".

---

*[src/core/ext/client_channel/client_channel.c, line 971 at r1](https://reviewable.io:443/reviews/grpc/grpc/9660#-Kcd2k4MIbTU_xuMFhTb:-Kcd2k4MIbTU_xuMFhTc:b-5u1znr) ([raw file](https://github.com/grpc/grpc/blob/3fc3408cd33eb2e221863e0881ec82eb8e21414f/src/core/ext/client_channel/client_channel.c#L971)):*
> ```C
> // initial metadata op, and (b) the need for efficiency, especially in
> // the streaming case.
> // TODO(ctiller): Explain this more thoroughly.
> ```

Might be a good time to address this TODO, as long as you're in here.

---

*[src/core/ext/client_channel/client_channel.c, line 1201 at r1](https://reviewable.io:443/reviews/grpc/grpc/9660#-Kcd5CwkxBidqpkgSBVY:-Kcd5CwkxBidqpkgSBVZ:b6jjdta) ([raw file](https://github.com/grpc/grpc/blob/3fc3408cd33eb2e221863e0881ec82eb8e21414f/src/core/ext/client_channel/client_channel.c#L1201)):*
> ```C
>   channel_data *chand = elem->channel_data;
>   grpc_connectivity_state out;
>   out = grpc_connectivity_state_check(&chand->state_tracker);
> ```

Can combine this with the previous line.

---

*[src/core/ext/client_channel/client_channel.c, line 1206 at r1](https://reviewable.io:443/reviews/grpc/grpc/9660#-Kcd5PzEmvd3dBSwjdX5:-Kcd5PzEmvd3dBSwjdX6:b-pn0l9i) ([raw file](https://github.com/grpc/grpc/blob/3fc3408cd33eb2e221863e0881ec82eb8e21414f/src/core/ext/client_channel/client_channel.c#L1206)):*
> ```C
>     grpc_closure_sched(
>         exec_ctx,
>         grpc_closure_create(try_to_connect_locked, chand,
> ```

I notice that, due to this change, this function may now return before we've called grpc_lb_policy_exit_idle() or grpc_resolver_next().  But since both of those functions operate asynchronously anyway, I'm guessing that's okay, right?

---

*[src/core/ext/client_channel/client_channel.c, line 1233 at r1](https://reviewable.io:443/reviews/grpc/grpc/9660#-Kcd6IEx_d47bWNk9Kd5:-Kcd6IEy_5GCgTwEg5bX:b-6znw86) ([raw file](https://github.com/grpc/grpc/blob/3fc3408cd33eb2e221863e0881ec82eb8e21414f/src/core/ext/client_channel/client_channel.c#L1233)):*
> ```C
> }
> 
> static void cc_watch_connectivity_state_locked(grpc_exec_ctx *exec_ctx,
> ```

The convention I've been trying to hold to in this code is that the "cc_" prefix is used only for methods that go in the filter's vtable.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/grpc/grpc/9660#-:-Kcd8lwsfraW4cE4tfgq:b-677jey)*
<!-- Sent from Reviewable.io -->

ctiller commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[src/core/ext/client_channel/client_channel.c, line 427 at r1](https://reviewable.io:443/reviews/grpc/grpc/9660#-KcczbwpRpQCU_BjKzgU:-KcxhjUSLR3SWTzQhhmW:b6js0w7) ([raw file](https://github.com/grpc/grpc/blob/3fc3408cd33eb2e221863e0881ec82eb8e21414f/src/core/ext/client_channel/client_channel.c#L427)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

It seems a bit messy to use transport private data here, since this is not transport code.  Could we instead create some sort of local data structure to pass in both op and elem via args?
</blockquote></details>

It's not possible to use a local data structure (possibly unbounded number of ops coming through), and we need to avoid the allocation.

I'll rename transport_private to handler_private instead.

---

*[src/core/ext/client_channel/client_channel.c, line 876 at r1](https://reviewable.io:443/reviews/grpc/grpc/9660#-Kcd0XSCchVKu8_DRk-4:-Kcxn-mWbEeAq_tftxT-:b-xku6h3) ([raw file](https://github.com/grpc/grpc/blob/3fc3408cd33eb2e221863e0881ec82eb8e21414f/src/core/ext/client_channel/client_channel.c#L876)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Same comment here as above -- I'd prefer to use some other mechanism to pass both op and elem in via args.
</blockquote></details>

It needs to be this or we incur all sorts of overhead - but I've updated the field name to be more obvious that this is ok.

---

*[src/core/ext/client_channel/client_channel.c, line 881 at r1](https://reviewable.io:443/reviews/grpc/grpc/9660#-Kcd1xz3nTvQgyjCQKwN:-KcxlDcpnISFfuucPY_O:b-jqlj71) ([raw file](https://github.com/grpc/grpc/blob/3fc3408cd33eb2e221863e0881ec82eb8e21414f/src/core/ext/client_channel/client_channel.c#L881)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Now that the logic here is a bit simpler, would it make sense to use a loop instead of a goto for this?

Feel free to push back on this -- I'm honestly not sure whether or not this would improve readability.  I do generally prefer loops to gotos, but in this particular case, it might be that "goto retry" and "goto done" are a bit more self-documenting than "continue" and "break".
</blockquote></details>

I moved the done branch into its own function (so we can just return).
The remaining retry cases I moved to a recursion... it's 50/50 to me whether that part reads better, but it'll be super useful when debugging...

---

*[src/core/ext/client_channel/client_channel.c, line 971 at r1](https://reviewable.io:443/reviews/grpc/grpc/9660#-Kcd2k4MIbTU_xuMFhTb:-KcxmBNAv5Gw_N2Daze_:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/3fc3408cd33eb2e221863e0881ec82eb8e21414f/src/core/ext/client_channel/client_channel.c#L971)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Might be a good time to address this TODO, as long as you're in here.
</blockquote></details>

Done.

---

*[src/core/ext/client_channel/client_channel.c, line 1201 at r1](https://reviewable.io:443/reviews/grpc/grpc/9660#-Kcd5CwkxBidqpkgSBVY:-KcxmLoJcapeI16Z2dTh:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/3fc3408cd33eb2e221863e0881ec82eb8e21414f/src/core/ext/client_channel/client_channel.c#L1201)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Can combine this with the previous line.
</blockquote></details>

Done.

---

*[src/core/ext/client_channel/client_channel.c, line 1206 at r1](https://reviewable.io:443/reviews/grpc/grpc/9660#-Kcd5PzEmvd3dBSwjdX5:-KcxmGTNwTNT48qIB0wW:b-vmmom0) ([raw file](https://github.com/grpc/grpc/blob/3fc3408cd33eb2e221863e0881ec82eb8e21414f/src/core/ext/client_channel/client_channel.c#L1206)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

I notice that, due to this change, this function may now return before we've called grpc_lb_policy_exit_idle() or grpc_resolver_next().  But since both of those functions operate asynchronously anyway, I'm guessing that's okay, right?
</blockquote></details>

I think so.

---

*[src/core/ext/client_channel/client_channel.c, line 1233 at r1](https://reviewable.io:443/reviews/grpc/grpc/9660#-Kcd6IEx_d47bWNk9Kd5:-KcxmbzOIlhgSiM3o66f:b-om39e2) ([raw file](https://github.com/grpc/grpc/blob/3fc3408cd33eb2e221863e0881ec82eb8e21414f/src/core/ext/client_channel/client_channel.c#L1233)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

The convention I've been trying to hold to in this code is that the "cc_" prefix is used only for methods that go in the filter's vtable.
</blockquote></details>

I've tried to update new functions as appropriate

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/grpc/grpc/9660)*
<!-- Sent from Reviewable.io -->

ctiller commented:
@dgquintas: can you finish up this review since @markdroth is out?

dgquintas commented:
On it

ctiller commented:
How's this one looking?

dgquintas commented on src/core/lib/iomgr/ev_poll_posix.c:
```diff
@@ -149,7 +149,7 @@ static void fd_end_poll(grpc_exec_ctx *exec_ctx, grpc_fd_watcher *rec,
 
 /* Reference counting for fds */
-/*#define GRPC_FD_REF_COUNT_DEBUG*/
+//#define GRPC_FD_REF_COUNT_DEBUG
```
nit: we are now mixing comment styles on this file.

ctiller commented:
Failed tests: #9550

markdroth commented on src/core/ext/client_channel/client_channel.c:
```diff
@@ -426,20 +421,12 @@ static void on_resolver_result_changed(grpc_exec_ctx *exec_ctx, void *arg,
+static void start_transport_op_locked(grpc_exec_ctx *exec_ctx, void *arg,
+                                      grpc_error *error_ignored) {
+  grpc_transport_op *op = arg;
+  grpc_channel_element *elem = op->transport_private.args[0];
```
It looks like you changed transport_private to handler_private in the stream op structure, but not in the transport op structure.
